File: backend/app/schemas/investment.py
"""Pydantic schemas for Investment endpoints."""

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.investment import AssetClass
from app.schemas._types import JSONDecimal

# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------


class InvestmentCreate(BaseModel):
    goal_id: uuid.UUID
    name: str = Field(..., min_length=1, max_length=200)
    asset_class: AssetClass
    expected_cagr: Decimal = Field(..., ge=0, le=100)
    notes: str | None = Field(None, max_length=1000)


class InvestmentUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=200)
    expected_cagr: Decimal | None = Field(None, ge=0, le=100)
    is_active: bool | None = None
    notes: str | None = Field(None, max_length=1000)


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------


class InvestmentResponse(BaseModel):
    id: uuid.UUID
    goal_id: uuid.UUID
    goal_name: str
    name: str
    asset_class: AssetClass
    expected_cagr: JSONDecimal
    is_active: bool
    # Computed from latest MonthlyEntry (None if no entries yet)
    latest_total_invested: JSONDecimal | None
    latest_current_value: JSONDecimal | None
    unrealized_gain: JSONDecimal | None
    absolute_return_pct: JSONDecimal | None
    latest_entry_month: date | None
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class InvestmentListResponse(BaseModel):
    investments: list[InvestmentResponse]
    count: int

File: frontend/src/types/investment.ts
import type { AssetClass } from "./goal";

export interface Investment {
  id: string;
  name: string;
  asset_class: AssetClass;
  expected_cagr: number;
  start_date: string;
  is_active: boolean;
  latest_total_invested: number;
  latest_current_value: number;
  unrealized_gain: number;
  absolute_return_pct: number;
  latest_entry_month: string | null;
  notes: string | null;
  linked_goals: LinkedGoal[];
  created_at: string;
  updated_at: string;
}

export interface LinkedGoal {
  goal_id: string;
  goal_name: string;
  allocation_pct: number;
}

export interface GoalInvestmentLink {
  id: string;
  goal_id: string;
  investment_id: string;
  allocation_pct: number;
  created_at: string;
}

export interface InvestmentCreate {
  name: string;
  asset_class: AssetClass;
  expected_cagr: number;
  start_date: string;
  notes?: string;
}

export interface InvestmentUpdate {
  name?: string;
  expected_cagr?: number;
  is_active?: boolean;
  notes?: string;
  start_date?: string;
}

export interface LinkInvestmentToGoal {
  investment_id: string;
  allocation_pct: number;
}
